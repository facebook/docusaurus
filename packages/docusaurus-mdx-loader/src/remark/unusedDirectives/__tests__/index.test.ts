/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import remark2rehype from 'remark-rehype';
import stringify from 'rehype-stringify';
import vfile from 'to-vfile';
import plugin from '../index';
import admonition from '../../admonitions';
import type {WebpackCompilerName} from '@docusaurus/utils';

const processFixture = async (
  name: string,
  {compilerName}: {compilerName: WebpackCompilerName},
) => {
  const {remark} = await import('remark');
  const {default: directives} = await import('remark-directive');

  const filePath = path.join(__dirname, '__fixtures__', `${name}.md`);
  const file = await vfile.read(filePath);
  file.data.compilerName = compilerName;

  const result = await remark()
    .use(directives)
    .use(admonition)
    .use(plugin)
    .use(remark2rehype)
    .use(stringify)
    .process(file);

  return result.value;
};

describe('directives remark plugin - client compiler', () => {
  const consoleMock = jest.spyOn(console, 'warn').mockImplementation(() => {});
  beforeEach(() => jest.clearAllMocks());

  const options = {compilerName: 'client'} as const;

  it('default behavior for container directives', async () => {
    const result = await processFixture('containerDirectives', options);
    expect(result).toMatchSnapshot('result');
    expect(consoleMock).toHaveBeenCalledTimes(1);
    expect(consoleMock.mock.calls).toMatchSnapshot('console');
  });

  it('default behavior for leaf directives', async () => {
    const result = await processFixture('leafDirectives', options);
    expect(result).toMatchSnapshot('result');
    expect(consoleMock).toHaveBeenCalledTimes(1);
    expect(consoleMock.mock.calls).toMatchSnapshot('console');
  });

  it('default behavior for text directives', async () => {
    const result = await processFixture('textDirectives', options);
    expect(result).toMatchSnapshot('result');
    expect(consoleMock).toHaveBeenCalledTimes(1);
    expect(consoleMock.mock.calls).toMatchSnapshot('console');
  });
});

describe('directives remark plugin - server compiler', () => {
  const consoleMock = jest.spyOn(console, 'warn').mockImplementation(() => {});
  beforeEach(() => jest.clearAllMocks());

  const options = {compilerName: 'server'} as const;

  it('default behavior for container directives', async () => {
    const result = await processFixture('containerDirectives', options);
    expect(result).toMatchSnapshot('result');
    expect(consoleMock).toHaveBeenCalledTimes(0);
  });

  it('default behavior for leaf directives', async () => {
    const result = await processFixture('leafDirectives', options);
    expect(result).toMatchSnapshot('result');
    expect(consoleMock).toHaveBeenCalledTimes(0);
  });

  it('default behavior for text directives', async () => {
    const result = await processFixture('textDirectives', options);
    expect(result).toMatchSnapshot('result');
    expect(consoleMock).toHaveBeenCalledTimes(0);
  });
});

describe('directives remark plugin - client result === server result', () => {
  // It is important that client/server outputs are exactly the same
  // otherwise React hydration mismatches can occur
  async function testSameResult(name: string) {
    const resultClient = await processFixture(name, {compilerName: 'client'});
    const resultServer = await processFixture(name, {compilerName: 'server'});
    expect(resultClient).toEqual(resultServer);
  }

  it('for containerDirectives', async () => {
    await testSameResult('containerDirectives');
  });

  it('for leafDirectives', async () => {
    await testSameResult('leafDirectives');
  });

  it('for textDirectives', async () => {
    await testSameResult('textDirectives');
  });
});
