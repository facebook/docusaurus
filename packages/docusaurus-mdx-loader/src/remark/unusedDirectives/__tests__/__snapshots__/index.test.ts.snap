// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`directives remark plugin - client compiler default behavior for container directives: console 1`] = `
[
  [
    "[WARNING] Docusaurus found 1 unused Markdown directives in file "packages/docusaurus-mdx-loader/src/remark/unusedDirectives/__tests__/__fixtures__/containerDirectives.md"
- :::unusedDirective (7:1)
Your content might render in an unexpected way. Visit https://github.com/facebook/docusaurus/pull/9394 to find out why and how to fix it.",
  ],
]
`;

exports[`directives remark plugin - client compiler default behavior for container directives: result 1`] = `
"<admonition type="danger"><p>Take care of snowstorms...</p></admonition>
<div><p>unused directive content</p></div>
<p>:::NotAContainerDirective with a phrase after</p>
<p>:::</p>
<p>Phrase before :::NotAContainerDirective</p>
<p>:::</p>"
`;

exports[`directives remark plugin - client compiler default behavior for leaf directives: console 1`] = `
[
  [
    "[WARNING] Docusaurus found 1 unused Markdown directives in file "packages/docusaurus-mdx-loader/src/remark/unusedDirectives/__tests__/__fixtures__/leafDirectives.md"
- ::unusedLeafDirective (1:1)
Your content might render in an unexpected way. Visit https://github.com/facebook/docusaurus/pull/9394 to find out why and how to fix it.",
  ],
]
`;

exports[`directives remark plugin - client compiler default behavior for leaf directives: result 1`] = `
"<div></div>
<p>Leaf directive in a phrase ::NotALeafDirective</p>
<p>::NotALeafDirective with a phrase after</p>"
`;

exports[`directives remark plugin - client compiler default behavior for text directives: console 1`] = `
[
  [
    "[WARNING] Docusaurus found 2 unused Markdown directives in file "packages/docusaurus-mdx-loader/src/remark/unusedDirectives/__tests__/__fixtures__/textDirectives.md"
- :textDirective3 (9:7)
- :textDirective4 (11:7)
Your content might render in an unexpected way. Visit https://github.com/facebook/docusaurus/pull/9394 to find out why and how to fix it.",
  ],
]
`;

exports[`directives remark plugin - client compiler default behavior for text directives: result 1`] = `
"<p>Simple: textDirective1</p>
<pre><code class="language-sh">Simple: textDirectiveCode
</code></pre>
<p>Simple:textDirective2</p>
<p>Simple<div>label</div></p>
<p>Simple<div></div></p>
<p>Simple:textDirective5</p>
<pre><code class="language-sh">Simple:textDirectiveCode
</code></pre>"
`;

exports[`directives remark plugin - server compiler default behavior for container directives: result 1`] = `
"<admonition type="danger"><p>Take care of snowstorms...</p></admonition>
<div><p>unused directive content</p></div>
<p>:::NotAContainerDirective with a phrase after</p>
<p>:::</p>
<p>Phrase before :::NotAContainerDirective</p>
<p>:::</p>"
`;

exports[`directives remark plugin - server compiler default behavior for leaf directives: result 1`] = `
"<div></div>
<p>Leaf directive in a phrase ::NotALeafDirective</p>
<p>::NotALeafDirective with a phrase after</p>"
`;

exports[`directives remark plugin - server compiler default behavior for text directives: result 1`] = `
"<p>Simple: textDirective1</p>
<pre><code class="language-sh">Simple: textDirectiveCode
</code></pre>
<p>Simple:textDirective2</p>
<p>Simple<div>label</div></p>
<p>Simple<div></div></p>
<p>Simple:textDirective5</p>
<pre><code class="language-sh">Simple:textDirectiveCode
</code></pre>"
`;
